### PR TITLE
Merge open-source branch 'fs_journal' into 'master'

### PR DESCRIPTION
**Pull request by amogh-samsung**
- New smartfs journaling logic implemented.
- New smartfs recovery logic implemented.
- Refactored smartfs fs layer operations for efficiency and clarity.
- New logic for smartfs_rename.
- Support renaming of directories also in addition to files.
- Optimization of parameters to smartfs fs layer functions.
- New smartfs auto powercut test app.
- Increase reliability and robustness of smartfs while reducing
  time and sector consumption in operations.
- Add truncate functionality to vfs and all file systems.
 (Changes uptil PR#4471)

Signed-off-by: Amogh Hassija a.hassija@samsung.com

### CODE REVIEW

**Review by Taejun-Kwon — APPROVED** [review]

**Comment by Taejun-Kwon** [comment]
Please check conflict first.

**Comment by sunghan-chang** [comment]
Instead of making your private branch and making merge commit, let's make a pr from fs_jounal branch directly.

**Comment by sunghan-chang** [comment]
Then, you can find below.
```
Conflicting files
build/configs/artik053/net-protected/defconfig
build/configs/imxrt1020-evk/tc_protected/defconfig
build/configs/imxrt1050-evk/hello_protected/defconfig
build/configs/imxrt1050-evk/helloxx_protected/defconfig
build/configs/imxrt1050-evk/tc_protected/defconfig
```

File: lib/libc/unistd/lib_truncate.c [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+ *      - A component of the path prefix of path is not a directory.
+ *    EROFS
+ *      - he named file resides on a read-only file system.
+ *    ENAMETOOLONG
+ *      - Pathname resolution of a symbolic link produced an intermediate
+ *        result whose length exceeds PATH_MAX.
+ *
+ ****************************************************************************/
+
+int truncate(FAR const char *path, off_t length)
```
Comment (thapav) [inline lib/libc/unistd/lib_truncate.c:129]: Can we split it into two commits ?
1) lib/libc/unistd: Add truncate API
2) os/fs: Add ftruncate to vfs & all file systems

File: os/fs/smartfs/Kconfig [hunk 2]
```diff
@@ -67,58 +67,17 @@ config SMARTFS_ALIGNED_ACCESS
...
 	default n
 	---help---
-		Enables recovery of lost sectors after power failure. Lost
-		sectors are the sectors which are allocated but not reachable
-		from root directory.
-
+		Instead of RTC, Use Time stamp for UTC value of entry.
 endmenu
-
+
```
Comment (thapav) [inline os/fs/smartfs/Kconfig:82]: Coding rule violation

File: lib/libc/unistd/lib_truncate.c (hunk 1 above)
Comment (Taejun-Kwon) [reply lib/libc/unistd/lib_truncate.c:129]: No it should be done by 1 commit, Or build error will be happened.

File: os/fs/smartfs/Kconfig (hunk 2 above)
Comment (Taejun-Kwon) [reply os/fs/smartfs/Kconfig:82]: I will modify later.

**Review by sunghan-chang — APPROVED** [review]

**Pull request merged**